src/Microsoft.AspNetCore.Http.Connections/HttpConnectionContext.cs

```diff
@@ -140,56 +140,86 @@ public void TickHeartbeat()
-                    if (ApplicationTask?.IsFaulted == true)
-                    {
-                        Transport?.Output.Complete(ApplicationTask.Exception.InnerException);
-                    }
-                    else
-                    {
-                        Transport?.Output.Complete();
-                    }
-
-                    // If the transport task is faulted, propagate the error to the application
-                    if (TransportTask?.IsFaulted == true)
-                    {
-                        Application?.Output.Complete(TransportTask.Exception.InnerException);
-                    }
-                    else
-                    {
-                        Application?.Output.Complete();
-                    }
-
                     var applicationTask = ApplicationTask ?? Task.CompletedTask;
                     var transportTask = TransportTask ?? Task.CompletedTask;
 
-                    disposeTask = WaitOnTasks(applicationTask, transportTask);
+                    disposeTask = WaitOnTasks(applicationTask, transportTask, closeGracefully);
                 }
             }
             finally
             {
                 Lock.Release();
             }
 
-            try
-            {
-                await disposeTask;
-            }
-            finally
-            {
-                // REVIEW: Should we move this to the read loops?
-
-                // Complete the reading side of the pipes
-                Application?.Input.Complete();
-                Transport?.Input.Complete();
-            }
+            await disposeTask;
         }
 
-        private async Task WaitOnTasks(Task applicationTask, Task transportTask)
+        private async Task WaitOnTasks(Task applicationTask, Task transportTask, bool closeGracefully)
         {
             try
             {
-                await Task.WhenAll(applicationTask, transportTask);
+                // Closing gracefully means we're only going to close the finished sides of the pipe
+                // If the application finishes, that means it's done with the transport pipe
+                // If the transport finishes, that means it's done with the application pipe
+                if (closeGracefully)
+                {
+                    // Wait for either to finish
+                    var result = await Task.WhenAny(applicationTask, transportTask);
+
+                    // If the application is complete, complete the transport pipe (it's the pipe to the transport)
+                    if (result == applicationTask)
+                    {
+                        Transport.Output.Complete(applicationTask.Exception?.InnerException);
+                        Transport.Input.Complete();
+
+                        try
+                        {
+                            // Transports are written by us and are well behaved, wait for them to drain
+                            await transportTask;
```

Comment:
Logging before and after this as well. Transports are more well-behaved, but that doesn't mean we can't make mistakes and would find value in seeing that a transport is hanging.